src/elfpy/pricing_models/yieldspace.py

```diff
@@ -56,42 +49,49 @@ def calc_lp_out_given_tokens_in(
+            "pricing_models.calc_lp_out_given_tokens_in: ERROR: "
+            f"Expected base_buffer >= 0, not {market_state.base_buffer}!"
+        )
+        assert market_state.lp_reserves >= 0, (
+            "pricing_models.calc_lp_out_given_tokens_in: ERROR: "
+            f"Expected lp_reserves >= 0, not {market_state.lp_reserves}!"
+        )
         assert rate >= 0, f"pricing_models.calc_lp_out_given_tokens_in: ERROR: expected rate >= 0, not {rate}!"
-        assert (
-            1 > time_remaining >= 0
-        ), f"pricing_models.calc_lp_out_given_tokens_in: ERROR: expected 1 > time_remaining >= 0, not {time_remaining}!"
-        assert stretched_time_remaining >= 0, (
+        assert 1 > time_remaining.normalized_time >= 0, (
+            "pricing_models.calc_lp_out_given_tokens_in: ERROR: "
+            f"expected 1 > time_remaining >= 0, not {time_remaining.normalied_time}!"
+        )
+        assert time_remaining.stretched_time >= 0, (
             "pricing_models.calc_lp_out_given_tokens_in: ERROR: "
-            f"expected stretched_time_remaining >= 0, not {stretched_time_remaining}!"
+            f"expected stretched_time_remaining >= 0, not {time_remaining.stretched_time}!"
         )
-        assert share_price >= init_share_price >= 1, (
+        assert market_state.share_price >= market_state.init_share_price >= 1, (
             "pricing_models.calc_lp_out_given_tokens_in: ERROR: "
             "expected share_price >= init_share_price >= 1, not "
-            f"share_price={share_price} and init_share_price={init_share_price}!"
+            f"share_price={market_state.share_price} and init_share_price={market_state.init_share_price}!"
         )
-        d_shares = d_base / share_price
-        if share_reserves > 0:  # normal case where we have some share reserves
+        d_shares = d_base / market_state.share_price
+        if market_state.share_reserves > 0:  # normal case where we have some share reserves
             # TODO: We need to update these LP calculations to address the LP
             #       exploit scenario.
-            lp_out = (d_shares * lp_reserves) / (share_reserves - base_buffer / share_price)
+            lp_out = (d_shares * market_state.lp_reserves) / (market_state.share_reserves - market_state.base_buffer)
         else:  # initial case where we have 0 share reserves or final case where it has been removed
             lp_out = d_shares
         # TODO: Move this calculation to a helper function.
-        d_bonds = (share_reserves + d_shares) / 2 * (
-            init_share_price * (1 + rate * time_remaining) ** (1 / stretched_time_remaining) - share_price
-        ) - bond_reserves
+        d_bonds = (market_state.share_reserves + d_shares) / 2 * (
+            market_state.init_share_price
+            * (1 + rate * time_remaining.normalized_time) ** (1 / time_remaining.stretched_time)
+            - market_state.share_price
+        ) - market_state.bond_reserves
```

Comment:
This logic should be abstracted (I realize that you didn't write it). The form is always the same except sometimes it's `market_state.share_reserves` or `market_state.share_reserves - d_shares`. Besides, having a helper function that computes the bond reserves given shares reserves, mu, c, and APR is super useful when writing unit tests.

Reply:
Yeah, there's a TODO for that right above the line